MachineLearning_Engine/Compute/ClimateCharts/UTCI.py

```diff
@@ -0,0 +1,272 @@
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3.0 of the License, or
+# (at your option) any later version.
+#
+# The BHoM is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
+#
+
+import os
+import numpy as np
+import pandas as pd
+import matplotlib
+matplotlib.use("agg")
+import matplotlib.pyplot as plt
+from matplotlib.colors import ListedColormap, BoundaryNorm
+from matplotlib.dates import date2num, DateFormatter
+
+ANNUAL_DATETIME = pd.date_range(start="2018-01-01 00:30:00", freq="60T", periods=8760, closed="left")
+
+
+def utci(
+    annual_values: np.ndarray, 
+    save_path: str, 
+    detailed: bool = False, 
+    title: str = None, 
+    tone_color: str = "black", 
+    invert_y: bool = False,
+    transparency: bool = False
+    ):
+    """ Generate a UTCI heatmap from a set of annual hourly values
+
+    Parameters
+    ----------
+    annual_values : np.ndarray
+        List of 8760 annual hourly values.
+    save_path : str
+        The full path where the plot will be saved.
+    detailed : bool
+        Set to True to include monthly stacked charts with proportion of time comfortable.
+    title : str or None
+        Adds a title to the plot.
+    tone_color : str or None
+        Text and border colours throughout the plot.
+    invert_y : bool
+        Reverse the y-axis so that 0-24 hours runs from top to bottom.
+    transparency : bool
+        Sets transparency of saved plot.
+
+    Returns
+    -------
+    imagePath : str
+        Path to the saved image
+    """
+
+    # Run input data checks
+    assert len(annual_values) == 8760, \
+        "Number of hourly values passed ({}) does not equal 8760".format(len(annual_values))
+
+    assert os.path.exists(os.path.dirname(save_path)), \
+        "\"{}\" does not exist".format(os.path.abspath(os.path.dirname(save_path)))
+
+    # Create DataFrame from passed values
+    df = pd.DataFrame(data=annual_values, index=ANNUAL_DATETIME, columns=[title])
+
+    # Reshape data into time/day matrix for heatmap plotting
+    annual_matrix = df.pivot_table(columns=df.index.date, index=df.index.time).values[::-1]
+
+    # Create UTCI colormap
+    utci_cmap = ListedColormap(['#0D104B', '#262972', '#3452A4', '#3C65AF', '#37BCED', '#2EB349', '#F38322', '#C31F25', '#7F1416', '#580002'])
+    utci_cmap_bounds = [-100, -40, -27, -13, 0, 9, 26, 32, 38, 46, 100]
+    utci_cmap_norm = BoundaryNorm(utci_cmap_bounds, utci_cmap.N)
+
+    bounds = np.arange(-41, 48, 1)
+    norm = BoundaryNorm(bounds, utci_cmap.N)
```

Comment:
This breaks - the number of colors should be at least the same number as `len(bounds)`
```suggestion
    norm = BoundaryNorm(bounds, max(len(bounds), utci_cmap.N))
```